std/assembly/string.ts

```diff
@@ -703,36 +706,38 @@ export namespace String {
       var str = __alloc(len << 1, idof<String>()); // max is one u16 char per u8 byte
       var strOff = str;
       while (bufOff < bufEnd) {
-        let cp = <u32>load<u8>(bufOff++);
-        if (cp < 128) {
-          if (nullTerminated && !cp) break;
-          store<u16>(strOff, cp);
-          strOff += 2;
-        } else if (cp > 191 && cp < 224) {
-          if (bufEnd - bufOff < 1) break;
-          store<u16>(strOff, (cp & 31) << 6 | load<u8>(bufOff++) & 63);
-          strOff += 2;
-        } else if (cp > 239 && cp < 365) {
-          if (bufEnd - bufOff < 3) break;
-          cp = (
-            (cp                  &  7) << 18 |
-            (load<u8>(bufOff)    & 63) << 12 |
-            (load<u8>(bufOff, 1) & 63) << 6  |
-             load<u8>(bufOff, 2) & 63
-          ) - 0x10000;
-          bufOff += 3;
-          store<u16>(strOff, 0xD800 | (cp >> 10));
-          store<u16>(strOff, 0xDC00 | (cp & 1023), 2);
-          strOff += 4;
+        let u0 = <u32>load<u8>(bufOff); ++bufOff;
+        if (!(u0 & 128)) {
+          // @ts-ignore: cast
+          if (nullTerminated & !u0) break;
+          store<u16>(strOff, u0);
         } else {
-          if (bufEnd - bufOff < 2) break;
-          store<u16>(strOff,
-            (cp                  & 15) << 12 |
-            (load<u8>(bufOff)    & 63) << 6  |
-             load<u8>(bufOff, 1) & 63
-          );
-          bufOff += 2; strOff += 2;
+          if (bufEnd == bufOff) break;
+          let u1 = <u32>load<u8>(bufOff) & 63; ++bufOff;
+          if ((u0 & 224) == 192) {
+            store<u16>(strOff, (u0 & 31) << 6 | u1);
+          } else {
+            if (bufEnd == bufOff) break;
+            let u2 = <u32>load<u8>(bufOff) & 63; ++bufOff;
+            if ((u0 & 240) == 224) {
+              u0 = (u0 & 15) << 12 | u1 << 6 | u2;
+            } else {
+              if (bufEnd == bufOff) break;
```

Comment:
On a first glimpse this seems to be trading branches, like where we previously checked that we can process 3 bytes (never emitting truncated bytes), we now check per byte (potentially emitting truncated bytes?). Makes me wonder what's better.